File: src/help_desk_pipeline/transformer.py
"""
Transformacja i walidacja danych ticketów / Ticket data transformation and validation.
Zaawansowana transformacja: walidacja, enrichment, filtrowanie, deduplikacja.
Advanced transformation: validation, enrichment, filtering, deduplication.
"""
import re
import hashlib
import logging
from datetime import datetime
from typing import List, Optional, Tuple

from models import (
    Ticket, TicketType, TicketPriority, TicketStatus,
    ConversationMessage, TransferResult
)

logger = logging.getLogger(__name__)


class TicketValidator:
    """
    Walidator ticketów / Ticket validator.
    Sprawdza czy ticket ma wszystkie wymagane dane.
    Checks if ticket has all required data.
    """

    REQUIRED_FIELDS = ["ticket_id", "ticket_type", "description"]

    @staticmethod
    def validate(ticket: Ticket) -> Tuple[bool, List[str]]:
        """
        Waliduj ticket / Validate ticket.

        Returns:
            (is_valid, list_of_errors)
        """
        errors = []

        # Sprawdź wymagane pola / Check required fields
        if not ticket.ticket_id:
            errors.append("Brak ticket_id / Missing ticket_id")
        if not ticket.description or not ticket.description.strip():
            errors.append("Brak opisu / Missing description")
        if ticket.ticket_type == TicketType.OTHER:
            logger.warning(f"Ticket {ticket.ticket_id} ma typ 'other' - może wymagać ręcznej kategoryzacji")

        # Walidacja konwersacji / Conversation validation
        if not ticket.conversation:
            errors.append("Brak wstępnej konwersacji / Missing initial conversation")

        for i, msg in enumerate(ticket.conversation):
            if not msg.content.strip():
                errors.append(f"Wiadomość #{i + 1} w konwersacji jest pusta / Message #{i + 1} is empty")
            if not msg.sender:
                errors.append(f"Wiadomość #{i + 1} nie ma nadawcy / Message #{i + 1} has no sender")

        is_valid = len(errors) == 0
        if not is_valid:
            logger.warning(f"Ticket {ticket.ticket_id} validation failed: {errors}")
        return is_valid, errors


class TicketEnricher:
    """
    Wzbogacanie danych ticketów / Ticket data enrichment.
    Automatyczna kategoryzacja i priorytetyzacja.
    Automatic categorization and prioritization.
    """

    # Słowa kluczowe do automatycznej kategoryzacji / Keywords for auto-categorization
    CATEGORY_KEYWORDS = {
        "billing": ["faktura", "płatność", "invoice", "payment", "billing", "rachunek", "cena"],
        "technical": ["błąd", "error", "crash", "awaria", "nie działa", "bug", "timeout", "500", "404"],
        "account": ["konto", "account", "logowanie", "login", "hasło", "password", "rejestracja"],
        "feature": ["proponuję", "sugestia", "feature", "request", "nowa funkcja", "dodaj"],
        "integration": ["api", "integracja", "webhook", "integration", "sdk", "połączenie"],
        "performance": ["wolno", "wydajność", "performance", "slow", "lag", "timeout"],
    }

    # Słowa kluczowe do priorytetyzacji / Keywords for prioritization
    PRIORITY_KEYWORDS = {
        TicketPriority.CRITICAL: [
            "krytyczny", "critical", "produkcja", "production", "awaria",
            "outage", "down", "nie działa", "emergency", "natychmiast"
        ],
        TicketPriority.HIGH: [
            "ważne", "important", "pilne", "urgent", "szybko",
            "asap", "blokujące", "blocking"
        ],
        TicketPriority.LOW: [
            "kiedyś", "someday", "niska", "low priority",
            "nie pilne", "not urgent", "pytanie", "question"
        ],
    }

    @classmethod
    def auto_categorize(cls, ticket: Ticket) -> str:
        """
        Automatycznie przypisz kategorię na podstawie opisu i konwersacji.
        Auto-assign category based on description and conversation.
        """
        text = ticket.description.lower()
        text += " " + " ".join(m.content.lower() for m in ticket.conversation)

        best_category = "general"
        best_score = 0

        for category, keywords in cls.CATEGORY_KEYWORDS.items():
            score = sum(1 for kw in keywords if kw in text)
            if score > best_score:
                best_score = score
                best_category = category

        logger.debug(f"Auto-categorized ticket {ticket.ticket_id}: {best_category} (score={best_score})")
        return best_category

    @classmethod
    def auto_prioritize(cls, ticket: Ticket) -> TicketPriority:
        """
        Automatycznie przypisz priorytet na podstawie treści.
        Auto-assign priority based on content.
        """
        text = ticket.description.lower()
        text += " " + " ".join(m.content.lower() for m in ticket.conversation)

        best_priority = TicketPriority.MEDIUM
        best_score = 0

        for priority, keywords in cls.PRIORITY_KEYWORDS.items():
            score = sum(1 for kw in keywords if kw in text)
            if score > best_score:
                best_score = score
                best_priority = priority

        logger.debug(f"Auto-prioritized ticket {ticket.ticket_id}: {best_priority.value} (score={best_score})")
        return best_priority

    @classmethod
    def generate_tags(cls, ticket: Ticket) -> List[str]:
        """
        Wygeneruj tagi na podstawie typu i kategorii.
        Generate tags based on type and category.
        """
        tags = [ticket.ticket_type.value]
        if ticket.category:
            tags.append(ticket.category)
        if ticket.priority in (TicketPriority.HIGH, TicketPriority.CRITICAL):
            tags.append("needs_attention")
        if len(ticket.conversation) > 3:
            tags.append("multi_message")
        return list(set(tags))

    @classmethod
    def enrich(cls, ticket: Ticket, auto_categorize: bool = True,
               auto_priority: bool = True) -> Ticket:
        """
        Wzbogać ticket o dodatkowe dane.
        Enrich ticket with additional data.
        """
        if auto_categorize and not ticket.category:
            ticket.category = cls.auto_categorize(ticket)

        if auto_priority and ticket.priority == TicketPriority.MEDIUM:
            ticket.priority = cls.auto_prioritize(ticket)

        ticket.tags = cls.generate_tags(ticket)
        ticket.transferred_at = datetime.now().isoformat()

        return ticket


class TicketFilter:
    """
    Filtrowanie ticketów / Ticket filtering.
    """

    @staticmethod
    def filter_tickets(
            tickets: List[Ticket],
            filter_resolved: bool = True,
            filter_closed: bool = True,
            exclude_ids: Optional[set] = None,
    ) -> List[Ticket]:
        """
        Filtruj tickety przed transferem.
        Filter tickets before transfer.
        """
        filtered = []
        excluded_count = 0

        for ticket in tickets:
            # Filtruj po statusie / Filter by status
            if filter_resolved and ticket.status == TicketStatus.RESOLVED:
                excluded_count += 1
                continue
            if filter_closed and ticket.status == TicketStatus.CLOSED:
                excluded_count += 1
                continue

            # Filtruj po ID (już przeniesione) / Filter by ID (already transferred)
            if exclude_ids and ticket.ticket_id in exclude_ids:
                excluded_count += 1
                continue

            filtered.append(ticket)

        logger.info(
            f"Filtered tickets: {len(filtered)} to transfer, "
            f"{excluded_count} excluded (resolved/closed/duplicate)"
        )
        return filtered


class TicketTransformer:
    """
    Główny klasa transformująca / Main transformation class.
    Łączy walidację, enrichment i filtrowanie.
    Combines validation, enrichment and filtering.
    """

    def __init__(self, auto_categorize: bool = True, auto_priority: bool = True):
        self.auto_categorize = auto_categorize
        self.auto_priority = auto_priority
        self.validator = TicketValidator()
        self.enricher = TicketEnricher()
        self.filter = TicketFilter()

    def process_ticket(self, ticket: Ticket) -> Tuple[Optional[Ticket], List[str]]:
        """
        Przetwórz pojedynczy ticket: walidacja + enrichment.
        Process single ticket: validation + enrichment.

        Returns:
            (processed_ticket or None, list_of_errors)
        """
        # 1. Walidacja / Validation
        is_valid, errors = self.validator.validate(ticket)
        if not is_valid:
            return None, errors

        # 2. Enrichment
        ticket = self.enricher.enrich(
            ticket,
            auto_categorize=self.auto_categorize,
            auto_priority=self.auto_priority,
        )

        # 3. Normalizacja danych / Data normalization
        ticket = self._normalize(ticket)

        return ticket, []

    def _normalize(self, ticket: Ticket) -> Ticket:
        """
        Normalizuj dane ticketa / Normalize ticket data.
        """
        # Oczyść opis / Clean description
        if ticket.description:
            ticket.description = ticket.description.strip()

        # Sortuj konwersację po czasie / Sort conversation by time
        ticket.conversation.sort(key=lambda m: m.timestamp)

        # Normalizuj nazwę klienta / Normalize client name
        if ticket.client_name:
            ticket.client_name = ticket.client_name.strip().title()

        return ticket

    def process_batch(
            self,
            tickets: List[Ticket],
            filter_resolved: bool = True,
            filter_closed: bool = True,
            exclude_ids: Optional[set] = None,
    ) -> Tuple[List[Ticket], List[TransferResult]]:
        """
        Przetwórz batch ticketów / Process batch of tickets.

        Returns:
            (list_of_valid_tickets, list_of_failed_results)
        """
        # 1. Filtrowanie / Filtering
        filtered = self.filter.filter_tickets(
            tickets,
            filter_resolved=filter_resolved,
            filter_closed=filter_closed,
            exclude_ids=exclude_ids,
        )

        # 2. Transformacja każdego ticketa / Transform each ticket
        valid_tickets = []
        failed_results = []

        for ticket in filtered:
            processed, errors = self.process_ticket(ticket)
            if processed:
                valid_tickets.append(processed)
            else:
                failed_results.append(TransferResult(
                    ticket_id=ticket.ticket_id,
                    success=False,
                    error_message="; ".join(errors),
                    timestamp=datetime.now().isoformat(),
                ))

        logger.info(
            f"Batch processing: {len(tickets)} input → "
            f"{len(valid_tickets)} valid, {len(failed_results)} failed"
        )
        return valid_tickets, failed_results

File: src/transformer.rs
use std::collections::HashSet;

use lazy_static::lazy_static;
use regex::Regex;
use tracing::{debug, info, warn};

use crate::models::{Ticket, TicketPriority, TicketStatus, TicketType, TransferResult};
use crate::security::{AuditLogger, ChecksumGenerator};

lazy_static! {
    static ref CATEGORY_KEYWORDS: Vec<(&'static str, Vec<&'static str>)> = vec![
        ("billing", vec!["faktura", "platnosc", "invoice", "payment", "billing", "rachunek", "cena"]),
        ("technical", vec!["blad", "error", "crash", "awaria", "nie dziala", "bug", "timeout", "500", "404"]),
        ("account", vec!["konto", "account", "logowanie", "login", "haslo", "password", "rejestracja"]),
        ("feature", vec!["proponuje", "sugestia", "feature", "request", "nowa funkcja", "dodaj"]),
        ("integration", vec!["api", "integracja", "webhook", "integration", "sdk", "polaczenie"]),
        ("performance", vec!["wolno", "wydajnosc", "performance", "slow", "lag", "timeout"]),
    ];

    static ref PRIORITY_KEYWORDS: Vec<(TicketPriority, Vec<&'static str>)> = vec![
        (TicketPriority::Critical, vec![
            "critical", "production", "outage", "down", "emergency"
        ]),
        (TicketPriority::High, vec![
            "important", "urgent", "asap", "blocking"
        ]),
        (TicketPriority::Low, vec![
            "someday", "low priority", "not urgent", "question"
        ]),
    ];

    static ref SENSITIVE_PATTERNS: Vec<Regex> = vec![
        Regex::new(r"\b\d{4}[-\s]?\d{4}[-\s]?\d{4}[-\s]?\d{4}\b").unwrap(),
        Regex::new(r"\b[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Z|a-z]{2,}\b").unwrap(),
        Regex::new(r"\b\d{3}[-.]?\d{3}[-.]?\d{4}\b").unwrap(),
    ];
}

pub fn auto_categorize(ticket: &Ticket) -> String {
    let text = build_search_text(ticket);
    let text_lower = text.to_lowercase();

    let mut best_category = "general".to_string();
    let mut best_score = 0usize;

    for (category, keywords) in CATEGORY_KEYWORDS.iter() {
        let score = keywords.iter().filter(|kw| text_lower.contains(kw)).count();
        if score > best_score {
            best_score = score;
            best_category = category.to_string();
        }
    }

    debug!("Auto-categorized ticket {}: {} (score={})", ticket.ticket_id, best_category, best_score);
    best_category
}

pub fn auto_prioritize(ticket: &Ticket) -> TicketPriority {
    let text = build_search_text(ticket);
    let text_lower = text.to_lowercase();

    let mut best_priority = TicketPriority::Medium;
    let mut best_score = 0usize;

    for (priority, keywords) in PRIORITY_KEYWORDS.iter() {
        let score = keywords.iter().filter(|kw| text_lower.contains(kw)).count();
        if score > best_score {
            best_score = score;
            best_priority = priority.clone();
        }
    }

    debug!("Auto-prioritized ticket {}: {:?} (score={})", ticket.ticket_id, best_priority, best_score);
    best_priority
}

pub fn generate_tags(ticket: &Ticket) -> Vec<String> {
    let mut tags: HashSet<String> = HashSet::new();
    tags.insert(ticket.ticket_type.to_string());

    if !ticket.category.is_empty() {
        tags.insert(ticket.category.clone());
    }

    if matches!(ticket.priority, TicketPriority::High | TicketPriority::Critical) {
        tags.insert("needs_attention".to_string());
    }

    if ticket.conversation.len() > 3 {
        tags.insert("multi_message".to_string());
    }

    let text = build_search_text(ticket);
    for pattern in SENSITIVE_PATTERNS.iter() {
        if pattern.is_match(&text) {
            tags.insert("contains_sensitive_data".to_string());
            break;
        }
    }

    tags.into_iter().collect()
}

pub fn enrich_ticket(
    ticket: &mut Ticket,
    auto_categorize_flag: bool,
    auto_priority_flag: bool,
    checksum_enabled: bool,
) {
    if auto_categorize_flag && ticket.category.is_empty() {
        ticket.category = auto_categorize(ticket);
    }

    if auto_priority_flag && matches!(ticket.priority, TicketPriority::Medium) {
        ticket.priority = auto_prioritize(ticket);
    }

    ticket.tags = generate_tags(ticket);
    ticket.transferred_at = Some(chrono::Utc::now());

    if checksum_enabled {
        ticket.checksum = Some(ChecksumGenerator::generate_ticket_checksum(ticket));
    }
}

pub fn filter_tickets(
    tickets: Vec<Ticket>,
    filter_resolved: bool,
    filter_closed: bool,
    exclude_ids: &HashSet<String>,
) -> (Vec<Ticket>, usize) {
    let initial_count = tickets.len();

    let filtered: Vec<Ticket> = tickets
        .into_iter()
        .filter(|t| {
            if filter_resolved && matches!(t.status, TicketStatus::Resolved) {
                return false;
            }
            if filter_closed && matches!(t.status, TicketStatus::Closed) {
                return false;
            }
            if exclude_ids.contains(&t.ticket_id) {
                return false;
            }
            true
        })
        .collect();

    let excluded_count = initial_count - filtered.len();
    info!(
        "Filtered tickets: {} to transfer, {} excluded (resolved/closed/duplicate)",
        filtered.len(),
        excluded_count
    );

    (filtered, excluded_count)
}

pub fn process_batch(
    tickets: Vec<Ticket>,
    auto_categorize_flag: bool,
    auto_priority_flag: bool,
    filter_resolved: bool,
    filter_closed: bool,
    exclude_ids: &HashSet<String>,
    checksum_enabled: bool,
    audit: &AuditLogger,
) -> (Vec<Ticket>, Vec<TransferResult>) {
    let (filtered, _) = filter_tickets(tickets, filter_resolved, filter_closed, exclude_ids);

    let mut valid_tickets = Vec::new();
    let mut failed_results = Vec::new();

    for mut ticket in filtered {
        match ticket.validate() {
            Ok(()) => {
                ticket.normalize();
                enrich_ticket(&mut ticket, auto_categorize_flag, auto_priority_flag, checksum_enabled);
                valid_tickets.push(ticket);
            }
            Err(errors) => {
                warn!("Ticket {} validation failed: {:?}", ticket.ticket_id, errors);
                audit.log_validation_error(&ticket.ticket_id, &errors);
                failed_results.push(TransferResult {
                    ticket_id: ticket.ticket_id,
                    success: false,
                    helpcenter_ticket_id: None,
                    error_message: Some(errors.join("; ")),
                    timestamp: chrono::Utc::now(),
                    retry_count: 0,
                });
            }
        }
    }

    info!(
        "Batch processing: {} valid, {} failed",
        valid_tickets.len(),
        failed_results.len()
    );

    (valid_tickets, failed_results)
}

fn build_search_text(ticket: &Ticket) -> String {
    let mut text = ticket.description.clone();
    for msg in &ticket.conversation {
        text.push(' ');
        text.push_str(&msg.content);
    }
    text
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::models::{ConversationMessage, MessageRole};

    fn make_test_ticket() -> Ticket {
        let mut ticket = Ticket::new("T-001", TicketType::Bug, "Application crash");
        ticket.conversation.push(ConversationMessage::new(
            "John Smith",
            MessageRole::Client,
            "Cannot log in, critical error",
        ));
        ticket
    }

    #[test]
    fn test_auto_categorize_technical() {
        let ticket = make_test_ticket();
        let category = auto_categorize(&ticket);
        assert_eq!(category, "technical");
    }

    #[test]
    fn test_auto_categorize_billing() {
        let mut ticket = Ticket::new("T-002", TicketType::Question, "Problem with invoice");
        ticket.conversation.push(ConversationMessage::new(
            "Client",
            MessageRole::Client,
            "Payment failed, invoice not paid",
        ));
        let category = auto_categorize(&ticket);
        assert_eq!(category, "billing");
    }

    #[test]
    fn test_auto_prioritize_critical() {
        let ticket = Ticket::new("T-003", TicketType::Bug, "Critical error in production");
        let priority = auto_prioritize(&ticket);
        assert_eq!(priority, TicketPriority::Critical);
    }

    #[test]
    fn test_ticket_validation() {
        let ticket = make_test_ticket();
        assert!(ticket.validate().is_ok());

        let invalid = Ticket::new("", TicketType::Other, "");
        let errors = invalid.validate().unwrap_err();
        assert!(!errors.is_empty());
    }

    #[test]
    fn test_filter_resolved() {
        let mut tickets = vec![make_test_ticket()];
        let mut resolved = make_test_ticket();
        resolved.status = TicketStatus::Resolved;
        tickets.push(resolved);

        let exclude = HashSet::new();
        let (filtered, excluded) = filter_tickets(tickets, true, true, &exclude);
        assert_eq!(filtered.len(), 1);
        assert_eq!(excluded, 1);
    }

    #[test]
    fn test_generate_tags() {
        let ticket = make_test_ticket();
        let tags = generate_tags(&ticket);
        assert!(tags.contains(&"bug".to_string()));
    }
}
